lib/miq_expression/tag.rb

```diff
@@ -1,17 +1,27 @@
-class MiqExpression::Tag
-  def self.parse(tag)
-    klass, ns = tag.split(".")
-    klass, ns = "nil", klass if ns.nil? # support managed-label
-    ns = "/" + ns.split("-").join("/")
-    ns = ns.sub(/(\/user_tag\/)/, "/user/") # replace with correct namespace for user tags
-    new(klass.safe_constantize, ns)
+class MiqExpression::Tag < MiqExpression::Field
+  TAG_REGEX = /
+(?<model_name>([[:alnum:]]*(::)?)+)
+\.(?<associations>([a-z_]+\.)*)
+(?<namespace>\bmanaged|user_tag\b)
+-(?<column>[a-z]+[_[:alnum:]]+)
+/x
+
+  ParseError = Class.new(StandardError)
+
+  def self.parse(field)
+    match = TAG_REGEX.match(field) || return
+
+    associations = match[:associations].split(".")
+    namespace = match[:namespace] == 'user_tag' ? 'user' : match[:namespace]
+    new(match[:model_name].classify.safe_constantize, "/#{namespace}/#{match[:column]}", associations)
   end
 
-  attr_reader :model, :namespace
+  attr_reader :model, :namespace, :associations
 
-  def initialize(model, namespace)
+  def initialize(model, namespace, associations = nil)
```

Comment:
It looks like we're always passing a third thing here - does it need to be optional?